src/components/Modal/Modal.tsx

```diff
@@ -0,0 +1,142 @@
+import * as classNames from 'classnames';
+import * as PropTypes from 'prop-types';
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+
+import Button, { ButtonProps } from '../Button/Button';
+import * as Rivet from '../Rivet';
+import ModalEvent from './ModalEvent';
+
+
+export interface ModalProps {
+    /**
+     * Determines whether the modal is shown or not
+     */
+    isOpen?: boolean;
+    /**
+     * Optional function to call to raise when the alert is dismissed. If undefined, the modal will act as a dialog (i.e., no close button, 
+     * will not close on outside click or escape key)
+     */
+    onDismiss?: Rivet.Action;
+    /**
+     * The content of the modal's header
+     */
+    title: React.ReactNode | string;
+    /**
+     * Button or buttons which will appear in the footer of the modal
+     */
+    controls: React.ReactElement<ButtonProps> | Array<React.ReactElement<ButtonProps>>;
+}
+
+
+class Modal extends React.PureComponent<ModalProps & React.HTMLAttributes<HTMLButtonElement>> {
+    
+    public static displayName = 'Modal';
+
+    public static propTypes = {
+        children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
+        className: PropTypes.string,
+        controls: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.shape({ type: PropTypes.oneOf([Button]) })), PropTypes.shape({ type: PropTypes.oneOf([Button]) })]).isRequired,
+        id: PropTypes.string,
+        isOpen: PropTypes.bool,
+        onDismiss: PropTypes.func,
+        title: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node, PropTypes.string]).isRequired
+    };
+
+    private eventHandler;
+
+    constructor(props) {
+        super(props);
+        this.eventHandler = ModalEvent.handler(this.handleClickOutside);
+    }
+
+    public componentDidMount() {
+        if (this.props.onDismiss) {
+            this.handleEventRegistration();
+        }
+    }
+
+    public componentDidUpdate(prevProps) {
+        if(this.props.onDismiss !== prevProps.onDismiss || this.props.isOpen !== prevProps.isOpen) {
+            this.handleEventRegistration();
+        }
+    }
+
+    public componentWillUnmount() {
+        this.eventHandler.deregister();
+    }    
+    
+    public render() {
+        const { children, className, controls, isOpen, onDismiss, title, id = Rivet.shortuid() } = this.props;
```

Comment:
This destructuring should be in alphabetical order and you might consider also spreading the remaining properties to pass down.